### ISSUE TITLE
Unwanted token burn incident in Nautilus wallet

### ISSUE THREAD

**Issue by greenhat** [ergoplatform/sigma-rust#590]
from discord https://discord.com/channels/668903786361651200/729692906209673257/995038064277737502

> A user reported on nautilus channel that he got a big amount of tokens burned while doing a regular transaction.

> After some investigation I found out that it could be some bug on sigma-rust. The wallet selected three unspent boxes as input with 275 distinct tokens in total and assembled the transaction using sigma-rust, for some reason the tokens that were supposed to included into the change box were not included and ended up getting burned.

> I managed to simulate the transaction and it looks like the issue is related with the limit of 255 tokens per box, if I manually remove some tokens from the inputs, the change tokens gets included normally into the change box. 

> Tested versions: v0.6.1 and v0.7.0

> Please, let me know if you need some additional clarification.

> Here is a test case simulating this issue: https://gist.github.com/capt-nemo429/16f59d6bd6f3e86209b2060e0cb685c9

> Transaction: https://explorer.ergoplatform.com/en/transactions/6f2e8307f74cec87566ae59e9b743c799e48c6645633f30d3c9df827bbb0c573 

> Here is the transaction generated by the code I've shared above: 

```

{
  "inputs": [
    {
      "boxId": "490148afdc36f5459bbfd84922a446abea9a1077e031822f377b0ff3a6e467e3",
      "extension": {}
    },
    {
      "boxId": "f0b6ff61c41692ddbb4a57acd8ff5ca7f113bbd89278c28535f8347b96279967",
      "extension": {}
    },
    {
      "boxId": "a9ca7b041922d1051c01196610b64eb61f17dbe1951a4e638620da7fba0a2855",
      "extension": {}
    }
  ],
  "dataInputs": [],
  "outputs": [
    {
      "value": 1000000,
      "ergoTree": "0008cd037fae590b9610a6d0f8634184b7dc4a76de8610a398492106de74b175e93a5412",
      "assets": [
        {
          "tokenId": "7d1568dce627141e6c9945328ef2bae5b7cc5c4f67fd873e288b808e68447db3",
          "amount": 1
        },
        {
          "tokenId": "45a773bb751f4a1cb652624026a054b1fd7e526896d9b566938398d9bbdf2716",
          "amount": 1
        },
        {
          "tokenId": "7d441ce250ffd0d70511899e5e9090f20a0443260983f76ef89a32cbb3eabda4",
          "amount": 1
        },
        {
          "tokenId": "e21d27d9bbfec0a385e33c8987f62d14955ade38b3030d97681868b7492c76ad",
          "amount": 1
        },
        {
          "tokenId": "1a77f5e1f185318c16c13402b4df133d439ab0e85382eab0b5598c7a8aa6a6c9",
          "amount": 1
        }
      ],
      "additionalRegisters": {},
      "creationHeight": 788658
    },
    {
      "value": 7765648,
      "ergoTree": "0008cd02f469fd785dbbc9112514655250c414d0c122d7df5eabc6966985b45b34542c02",
      "assets": [],
      "additionalRegisters": {},
      "creationHeight": 788658
    },
    {
      "value": 1100000,
      "ergoTree": "1005040004000e36100204a00b08cd0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798ea02d192a39a8cc7a701730073011001020402d19683030193a38cc7b2a57300000193c2b2a57301007473027303830108cdeeac93b1a57304",
      "assets": [],
      "additionalRegisters": {},
      "creationHeight": 788658
    }
  ]
}
```

**Comment by kettlebell** [comment]
I looked into this and I think I've narrowed down the cause. The problem is that `BoxSelection::change_boxes` is of type `pub type BoxTokens = BoundedVec<Token, 1, 255>`. When we have more than 255 tokens as input, we will accidentally burn the tokens due to the following lines: https://github.com/ergoplatform/sigma-rust/blob/26155df207dc88601de1ae06a5bf5b049d8dcabf/ergo-lib/src/wallet/box_selector/simple.rs#L155-L159

That final `.ok()` call will convert an error (which will happen because we have > 255 tokens) into a `None` instance, thus losing the tokens.

**Comment by kettlebell** [comment]
If we're dealing with a similar situation in the future, we may need change `TxBuilder` to allow for multiple change addresses to hold the remaining tokens?

**Comment by greenhat** [comment]
@kettlebell Good catch! I found it as well. `.ok()` in such cases becomes very dangerous. I'm going to review all `Result` to `Option` conversions in our codebase.

Besides that, I'm planning to add a hard requirement to the box selector that `inputs = target + change` as a guard to be sure that we don't accidentally burn any tokens. 

Multiple change boxes can be guarded with the same PK.
BTW I also found that min_change_value in TxBuilder could cause a change box(es) omitted. I'm thinking of removing it completely.
